src/backend/distributed/commands/foreign_server.c

```diff
@@ -0,0 +1,307 @@
+/*-------------------------------------------------------------------------
+ *
+ * foreign_server.c
+ *    Commands for FOREIGN SERVER statements.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/pg_foreign_server.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/commands.h"
+#include "distributed/deparser.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata/distobject.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/worker_transaction.h"
+#include "foreign/foreign.h"
+#include "nodes/primnodes.h"
+
+static Node * RecreateForeignServerStmt(Oid serverId);
+
+
+/*
+ * PreprocessCreateForeignServerStmt is called during the planning phase for
+ * CREATE SERVER.
+ */
+List *
+PreprocessCreateForeignServerStmt(Node *node, const char *queryString,
+								  ProcessUtilityContext processUtilityContext)
+{
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessAlterForeignServerStmt is called during the planning phase for
+ * ALTER SERVER .. OPTIONS ..
+ */
+List *
+PreprocessAlterForeignServerStmt(Node *node, const char *queryString,
+								 ProcessUtilityContext processUtilityContext)
+{
+	AlterForeignServerStmt *stmt = castNode(AlterForeignServerStmt, node);
+	ForeignServer *server = GetForeignServerByName(stmt->servername, false);
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
```

Comment:
Do we need `EnsureDependenciesExistOnAllNodes` here? I think with `AlterServer` you cannot alter the server to depend on a new object, right? Such as `ALTER SERVER .. SET SCHEMA new_schema` or such? If so, we should be fine.

As far as I can read from the docs, ALTER SERVER can only change Server's options?

Reply:
It's allowed to alter a server's owner, name and options. I thought like we might need to ensure the new owner role exists on the workers?

Reply:
Ah, that's why we do it on `PostProcess..` ?